throughput/spam.go

```diff
@@ -0,0 +1,427 @@
+	maxUnits, err = chain.EstimateUnits(parser.Rules(time.Now().UnixMilli()), actions, factory)
+	if err != nil {
+		return err
+	}
+	return s.returnFunds(ctx, cli, parser, maxUnits, sh, accounts, factories, funds, symbol)
+}
+
+func (s *Spammer) logZipf(zipfSeed *rand.Rand) {
+	zz := rand.NewZipf(zipfSeed, s.sZipf, s.vZipf, uint64(s.numAccounts)-1)
+	trials := s.txsPerSecond * 60 * 2 // sender/receiver
+	unique := set.NewSet[uint64](trials)
+	for i := 0; i < trials; i++ {
+		unique.Add(zz.Uint64())
+	}
+	utils.Outf("{{blue}}unique participants expected every 60s:{{/}} %d\n", unique.Len())
+}
+
+// createIssuer creates an [numClients] transaction issuers for each URI in [uris]
+func (s *Spammer) createIssuers(parser chain.Parser) ([]*issuer, error) {
+	issuers := []*issuer{}
+	for i := 0; i < len(s.uris); i++ {
+		for j := 0; j < s.numClients; j++ {
+			cli := jsonrpc.NewJSONRPCClient(s.uris[i])
+			webSocketClient, err := ws.NewWebSocketClient(s.uris[i], ws.DefaultHandshakeTimeout, pubsub.MaxPendingMessages, pubsub.MaxReadMessageSize) // we write the max read
+			if err != nil {
+				return nil, err
+			}
+			issuer := &issuer{i: len(issuers), cli: cli, ws: webSocketClient, parser: parser, uri: s.uris[i]}
+			issuers = append(issuers, issuer)
+		}
+	}
+	return issuers, nil
+}
+
+func (s *Spammer) distributeFunds(ctx context.Context, cli *jsonrpc.JSONRPCClient, parser chain.Parser, feePerTx uint64, sh SpamHelper) ([]*auth.PrivateKey, map[codec.Address]uint64, []chain.AuthFactory, error) {
+	withholding := feePerTx * uint64(s.numAccounts)
+	if s.balance < withholding {
+		return nil, nil, nil, fmt.Errorf("insufficient funds (have=%d need=%d)", s.balance, withholding)
+	}
+
+	distAmount := (s.balance - withholding) / uint64(s.numAccounts)
+
+	utils.Outf("{{yellow}}distributing funds to each account{{/}}\n")
+
+	funds := map[codec.Address]uint64{}
+	accounts := make([]*auth.PrivateKey, s.numAccounts)
+	factories := make([]chain.AuthFactory, s.numAccounts)
+
+	factory, err := auth.GetFactory(s.key)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	webSocketClient, err := ws.NewWebSocketClient(s.uris[0], ws.DefaultHandshakeTimeout, pubsub.MaxPendingMessages, pubsub.MaxReadMessageSize) // we write the max read
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	p := &pacer{ws: webSocketClient}
+	go p.Run(ctx, s.minTxsPerSecond)
+	// This helps from hanging
+	time.Sleep(3 * time.Second)
+	for i := 0; i < s.numAccounts; i++ {
+		// Create account
+		pk, err := sh.CreateAccount()
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		accounts[i] = pk
+		f, err := auth.GetFactory(pk)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		factories[i] = f
+
+		// Send funds
+		actions := sh.GetTransfer(pk.Address, distAmount, uniqueBytes())
+		// TODO: shouldn't this be using the factory from the account? rather the root key factory?
+		_, tx, err := cli.GenerateTransactionManual(parser, actions, factory, feePerTx)
+		if err != nil {
```

Comment:
We're using the root key factory to send funds from the rootKey to the newly created auth factories